src/EditorFeatures/Core.Wpf/Preview/PreviewFactoryService.cs

```diff
@@ -274,33 +274,38 @@ private Task<DifferenceViewerPreview> CreateAddedDocumentPreviewViewCoreAsync(IT
                 sourceSpans: new List<object> { firstLine, "\r\n", span }, registryService: _contentTypeRegistryService);
 
-            return CreateNewDifferenceViewerAsync(null, workspace, originalBuffer, changedBuffer, zoomLevel, cancellationToken);
+#pragma warning disable CA2007 // Consider calling ConfigureAwait on the awaited task (containing method uses JTF)
+            return await CreateNewDifferenceViewerAsync(null, workspace, originalBuffer, changedBuffer, zoomLevel, cancellationToken);
+#pragma warning restore CA2007 // Consider calling ConfigureAwait on the awaited task
         }
 
-        private Task<DifferenceViewerPreview> CreateAddedTextDocumentPreviewViewAsync(
+        private async Task<DifferenceViewerPreview> CreateAddedTextDocumentPreviewViewAsync(
             TextDocument document,
             double zoomLevel,
-            Func<TextDocument, CancellationToken, ITextBuffer> createBuffer,
+            Func<TextDocument, CancellationToken, ValueTask<ITextBuffer>> createBufferAsync,
             Action<Workspace, DocumentId> openTextDocument,
             CancellationToken cancellationToken)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-            var newBuffer = createBuffer(document, cancellationToken);
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var newBuffer = await createBufferAsync(document, cancellationToken);
 
             // Create PreviewWorkspace around the buffer to be displayed in the diff preview
             // so that all IDE services (colorizer, squiggles etc.) light up in this buffer.
             var rightWorkspace = new PreviewWorkspace(
                 document.Project.Solution.WithTextDocumentText(document.Id, newBuffer.AsTextContainer().CurrentText));
             openTextDocument(rightWorkspace, document.Id);
 
-            return CreateAddedDocumentPreviewViewCoreAsync(newBuffer, rightWorkspace, document, zoomLevel, cancellationToken);
+#pragma warning disable CA2007 // Consider calling ConfigureAwait on the awaited task (containing method uses JTF)
+            return await CreateAddedDocumentPreviewViewCoreAsync(newBuffer, rightWorkspace, document, zoomLevel, cancellationToken);
+#pragma warning restore CA2007 // Consider calling ConfigureAwait on the awaited task
         }
 
         public Task<DifferenceViewerPreview> CreateAddedDocumentPreviewViewAsync(Document document, double zoomLevel, CancellationToken cancellationToken)
         {
             return CreateAddedTextDocumentPreviewViewAsync(
                 document, zoomLevel,
-                createBuffer: (textDocument, cancellationToken) => CreateNewBuffer((Document)textDocument, cancellationToken),
+                createBufferAsync: (textDocument, cancellationToken) => CreateNewBufferAsync((Document)textDocument, cancellationToken),
```

Comment:
nit: these called functions and createBufferAsync could be generic on a TDocument type to prevent the needs for these casts.

Reply:
fixed